Quality of life changes to rest.nvim

udayvir-singh opened the pull request:
## Features
#### Add ability to add custom formatters.
setup function now accepts `formatters` table. this makes `jq` optional.

```lua
require [[rest-nvim]].setup {
  result = {
    -- executables for formatting response body [optional]
    formatters = {
      -- set them to nil if you want to disable them
      json = "jq",
      html = {"tidy", "-i", "-q", "--show-errors", "0", "-"}
    },
  }
}
```

#### Add better syntax highlights in httpResult.
- properly highlighted http status codes.
- add separate highlights for date/expires headers.
- better regex for numbers.
- support highlighting response based on content-type header.

![highlights](https://user-images.githubusercontent.com/97400310/183245876-8c49305b-2c28-4eee-99f2-d4e2a178e8ed.png)

#### Add support for passing flags to curl, fixes #123.

Example:
```js
POST http://localhost:3000/
Content-Type: application/json

-c /tmp/cookie

{ "id": 1 }
```

## Bug Fixes
#### Fix errors in plenary when parsing json, fixes #118 and fixes #114.

#### Fix request preview getting cropped off.

#### Fix #67 error handling for formatters.

NTBBloodbath commented on README.md:
```diff
@@ -47,7 +47,6 @@ have to leave Neovim!
 
 - System-wide
   - curl
-  - jq +1.6 (to format JSON output so it can be human-readable)
```
Maybe keep it as an optional dependency?

Also I think tidy should be added as an optional dependency too as AFAIK it is an external tool right?

udayvir-singh replied:
@NTBBloodbath

`jq` and `tidy` can be replaced by other executable or even native lua functions, that's why I didn't add them as dependencies.

udayvir-singh replied:
I don't think they should be added as optional dependencies as people would just install them thinking they are required.

NTBBloodbath replied:
Ah, makes sense now. Yeah let's leave it as this then, thanks for explaining

NTBBloodbath commented on lua/rest-nvim/request/init.lua:
```diff
@@ -42,7 +42,8 @@ end
 -- @param start_line Line where body starts
 -- @param stop_line Line where body stops
-local function get_body(bufnr, start_line, stop_line)
+-- @param has_hson True if content-type is set to json
```
`has_json` parameter name has a small typo here :p

NTBBloodbath reviewed (commented):
Hey, thank you so much for this!

All these changes looks great, just a small things to be fixed (only documentation) imho.

I'll hopefully merge this PR once these changes are made :)

Cheers

teto commented on lua/rest-nvim/curl/init.lua:
```diff
@@ -114,6 +147,19 @@ local function create_callback(method, url)
+
+    if vim.fn.filereadable(syntax_file) == 1 then
+      vim.cmd(string.gsub([[
+        unlet b:current_syntax
```
this line triggers an error "unknown b:current_syntax" if I run a request twice

NTBBloodbath replied:
I did push a small fix for this that checks if `b:current_syntax` exists a few seconds ago, please let me know if the issue persists.

teto replied:
seems like that fixed it thx

teto commented:
I've been trying to override/remove the default formatters since otherwise I get 
```
Error calling formatter on response body:                                                                                                                                                                                                                       
Vim:E475: Invalid value for argument cmd: 'tidy' is not executable
```

so I've done:
```
require("rest-nvim").setup({
          -- Open request results in a horizontal split
          result_split_horizontal = false,
          -- Skip SSL verification, useful for unknown certificates
          skip_ssl_verification = false,
          -- Highlight request on run
          highlight = {
		   enabled = true,
		   timeout = 150,
          },
          result = {
		   -- toggle showing URL, HTTP info, headers at top the of result window
		   show_url = true,
		   show_http_info = true,
		   show_headers = true,
		   -- disable formatters else they generate errors/add dependencies
		   -- for instance when it detects html, it tried to run 'tidy'
		   formatters = {
			html = nil,
			jq = nil
		   },
          },
          -- Jump to request line on run
          jump_to_request = false,
		})
```
to no avail. Can you change/override the formatters ? if yes how ? if no, do you want to send a patch ?

udayvir-singh commented:
@teto

I just noticed you have to set formatter to `false` instead of `nil` to disable them, also I have added an extra check to disable html formatter if tidy is not available in PR #165, and fixed the README in it.

teto commented:
do you know why we special case json when we send the payload ? there should be no need, shouldn't it ? 

I am looking at https://github.com/rest-nvim/rest.nvim/issues/157 and so if we didn't have to special case json, then we could skip the content-type check.
